### PR TITLE
`crux-mir`: Support `nightly-2023-01-23`

### PR DESCRIPTION
**Pull request by RyanGlScott**
This is a fairly large overhaul of `crux-mir`'s and `crucible-mir`'s internals to support a more recent Rust nightly (`nightly-2023-01-23`). The work is split between myself, @m10f, and @spernsteiner.

There is a lot happening here, so if you want to learn specifics, I'd encourage you to look at individual commits. Some of the broad themes in the patchset include:

* Propertly support for Rust's new constant forms
* Better support for zero-sized constants
* Encoding enum discriminant types so that `crucible-mir` can know about non-`isize` discriminant types (e.g., `Ordering`, which uses an `i8` discriminant)
* Being much more systematic about which patches we must apply to the Rust standard libraries for Crucible's needs (see the new `crux-mir/lib/Patches.md` file for explanations of each patch)
* A more intelligent way of computing crate disambiguators for looking up known types such as `MaybeUninit` and `Option`
* Re-enabling the macOS CI for `crux-mir`, which was previously disabled due to the nightly not working on recent macOSes

This should fix several issues:

* https://github.com/GaloisInc/crucible/issues/1050
* https://github.com/GaloisInc/crucible/issues/1059

### CODE REVIEW

File: crux-mir/lib/Patches.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  We also make sure to avoid the `Layout::array` function, which has a
+  particularly tricky use of `transmute` that we do not currently support.
+
+* Use crucible's allocator instead of `rustc_box` (last applied: May 17, 2023)
+
+  Same reasoning as above.
+
+* Reimplement `core::fmt` using `crucible::any::Any` (last applied: May 2, 2023)
+
+  TODO: Describe why this is necessary
```
Comment (RyanGlScott) [inline crux-mir/lib/Patches.md:29]: Oops, I left a `TODO` here. I should resolve this TODO before merging (which shouldn't be too hard).

File: crux-mir/lib/Patches.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+* Implement `HashMap` in terms of `Vec` (last applied: May 4, 2023)
+
+  TODO: Describe why this is necessary
```
Comment (RyanGlScott) [inline crux-mir/lib/Patches.md:33]: Resolve this `TODO`

**Review by spernsteiner — COMMENTED** [review]
Mostly looks good, aside from a possible correctness issue with the handling of `StatementKind::Intrinsic`.

File: crucible-mir/src/Mir/DefId.hs [hunk 3]
```diff
@@ -47,19 +52,21 @@ type Segment = (Text, Int)
...
 
+makeLenses ''DefId
+
 -- | The crate disambiguator hash produced when the crate metadata string is
--- empty.  This is the disambiguator for all sysroot crates, which are the only
--- ones we override.
+-- empty. This is useful for creating 'DefId's for thing that do not have a
+-- disambiguator associated with it (e.g., Crucible names).
 defaultDisambiguator :: Text
 defaultDisambiguator = "3a1fbbbh"
```
[the remark this answers is not in the document]
Comment (RyanGlScott) [reply crucible-mir/src/Mir/DefId.hs:71]: `defaultDisambiguator` is only used in `textId` when the provided string doesn't contain a disambiguator. Perhaps we should just `error` in that case instead and remove `defaultDisambiguator` entirely, however.

File: crucible-mir/src/Mir/Trans.hs [hunk 4]
```diff
@@ -978,11 +1034,15 @@ evalRval rv@(M.RAdtAg (M.AdtAg adt agv ops ty)) = do
...
-            M.Enum -> buildEnum adt (fromInteger agv) es
+            M.Enum _ -> buildEnum adt (fromInteger agv) es
             M.Union -> do
                 mirFail $ "evalRval: Union types are unsupported, for " ++ show (adt ^. adtname)
       _ -> mirFail $ "evalRval: unsupported type for AdtAg: " ++ show ty
+evalRval (M.ThreadLocalRef did _) = staticPlace did >>= addrOfPlace
 
+-- TODO: are these correct?
+evalRval rv@(M.ShallowInitBox op ty) = evalOperand op
+evalRval rv@(M.CopyForDeref lv) = evalLvalue lv
```
[the remark this answers is not in the document]
Comment (RyanGlScott) [reply crucible-mir/src/Mir/Trans.hs]: Thanks! I've added these explanations as comments.

File: crucible-mir/src/Mir/Trans.hs [hunk 5]
```diff
@@ -1162,6 +1233,8 @@ transStatement (M.SetDiscriminant lv i) = do
     -- simultaneously), then we could remove AllocateEnum.
     ty -> mirFail $ "don't know how to set discriminant of " ++ show ty
 transStatement M.Nop = return ()
+transStatement M.Deinit = return ()
+transStatement (M.StmtIntrinsic _) = return () --TODO: is this true?
```
[the remark this answers is not in the document]
Comment (RyanGlScott) [reply crucible-mir/src/Mir/Trans.hs]: I've pushed a commit which implements `NDIAssume` and `NDICopyNonOverlapping` properly. Happily, the test suite continues to work after these changes.

File: crucible-mir/src/Mir/DefId.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (RyanGlScott) [reply crucible-mir/src/Mir/DefId.hs]: I've pushed a commit that removes `defaultDisambiguator`, as well as any uses of `textId` that were implicitly relying on `defaultDisambiguator` being used. With that, I believe we have finally expunged all remaining uses.

**Comment by RyanGlScott** [comment]
It turns out that this broke the nightly `crux-mir` Docker job, as seen [here](https://github.com/GaloisInc/crucible/actions/runs/5574287177/jobs/10182706184). Thankfully, I was notified very quickly about this! I've submitted a fix in #1099.

**Pull request merged**